### ISSUE TITLE
Technical debt stand down

### ISSUE THREAD

**Issue by AdrianPotter** [ISISComputingGroup/IBEX#2817]
The intention of technical debt stand downs is that they are a day where we can prioritise a cluster of tickets that in themselves make a significant contribution to the overall quality of the project but individually would be unlikely to be prioritised. They are also an excellent opportunity for knowledge sharing within the team. 

The previous stand-down was #2725

Ahead of the day, we should:

- gather a list of potential topics and decide which would be of most value
- pick out tickets that would suitable to target and ideally assign a priority

There is a stand-down label that can be used for these tickets.

**Comment by AdrianPotter** [comment]
Ideas proposed via the G39 whiteboard:

- Documentation, @DominicOram 
- Bugs, me
- View models, @mattclarke 
- Unit tests, @DominicOram 
- IOC tests, anon
    - Integrate Lewis fully into IOC test framework, anon
    - "Finish rewrite of run_all_tests.py in IOC test framework", anon

I think anon above is @John-Holt-Tessella. I will invite proposers to make their case for each topic on Slack, followed by a poll. Simultaneously, I'll try and find a day that fits most people.

**Comment by AdrianPotter** [comment]
I've provisionally set the stand down for 15th Jan

**Comment by AdrianPotter** [comment]
Meeting invite sent. There is an ongoing discussion on Slack about what topic to pick, which requires input from @DominicOram who is away until Monday. I will park this in impeded until then.

**Comment by AdrianPotter** [comment]
The proposed topics for the stand down, and the justifications from their proposer are as follows.

# IOC Tests
It would be valuable to have some emulators/tests for some of our older devices (probably targeting commonly used kit). Even if the test suites aren't complete, it provides reassurance that the IOC launches correctly and that the major functionality is working correctly.

# Integrate Lewis fully into IOC test framework
Currently the IOC testing framework creates a new sub process and then launches the lewis emulator. This makes it hard to access some properties and functions on a device. Making it part of IOC testing framework would be useful.
** Should be combined with IOC Tests? **

# Bugs
We have 105 bug tickets in the backlog. Some of these have gone through several rounds of proposal but never get fixed owing to being individually too low priority. As a collection though, they contribute noticeably to the overall quality of the system. There's currently no other project mechanism by which these issues will ever be resolved. We can always focus on a particular class of bugs if we want to narrow the scope, but it would allow us to improve overall system quality in at least one area.

#2320 Improve EPICS build dependencies
Improve build speeds (possible parallelisation) and spread knowledge of the build system

# Documentation
Our documentation is, in some places, really very poor. We have a number of duplicate pages (search motors in the wiki), dead links and the overall structure is confusing in places. This should be a top priority as it is required for new starters, knowledge sharing and better collaboration. Documentation should of course be done whilst we work but I believe it requires a focus day to make it a good example for the future. I know this isn't the most exciting of topics but how many times have you had to struggle through understanding a bit of IBEX based on the code or having to find the right person to talk to?

# View Models
Issues in the GUI generally do not cause lost beam time but they are the things that affect our users the most. Many of the issues we have seen occur in the logic for how things are displayed e.g. buttons incorrectly greyed out or the synoptic failing to display on occasion. These issues are hard to fix, and impossible to test, as the logic for behaviour is mixed in with that for displaying. A concerted effort to get some of the more complex parts of the GUI into view models would mean that they are more reliable and testable.

# Unit Tests
Most people agree that unit tests make code more understandable, reliable and extensible.  We currently have a number of issues relating to our unit tests in IBEX:
- The amount of unit testing we have in IBEX is patchy, particularly in the GUI. 
- We have no metric on the coverage of our tests so do not really know if the components that have lots of tests are well tests.
- Many tests were written before we started using GIVEN_WHEN_THEN and so are vague in what they are testing.
- We have cases where unit tests are not run as part of the build, leaving them to be forgotten.
- We may still have cases where unit tests have external dependencies (as we recently had in the database server)

A concerted effort to fix some of these issues should mean that we can rely more on our tests going forward and developers are confident in writing them.

**Comment by AdrianPotter** [comment]
Awaiting poll feedback

**Comment by AdrianPotter** [comment]
The Slack poll has found slightly in favour of bug fixing. The results were:

Bugs, 5
Documentation, 4
IOC tests, 3
View models, 2
Others, 0

We should keep that in mind for future stand downs. In absence of a sensible reason to choose otherwise, we'll do bugs this stand down.

**Comment by AdrianPotter** [comment]
Put back in ready ahead of Monday...
